geps/gep-1713/index.md

```diff
@@ -0,0 +1,377 @@
+# GEP-1713: Standard Mechanism to Merge Multiple Gateways
+
+* Issue: [#1713](/kubernetes-sigs/gateway-api/issues/1713)
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## tl;dr
+
+The Gateway Resource is a contention point since it is the only place to attach listeners with certificates. We propose a mechanism to allow distinct Gateway resources to be logically merged.
+
+## Goals
+
+- Define a mechanic to merge multiple Gateways (logically)
+- Define a set of acceptable properties that can be merged and their semantics
+
+## Non-Goals
+
+- Apply a Gateway resource onto N distinct gateways (one to many)
+
+## Introduction
+
+Knative generates on demand per-service certificates using HTTP-01 challenges. 
+There can be O(1000) Knative Services in the cluster which means we have O(1000) distinct certificates. 
+Thus updating a single Gateway resource with this many certificates is a contention point and inhibits horizontal scaling of our controllers.
+[Istio Ambient](https://istio.io/v1.15/blog/2022/introducing-ambient-mesh/), similarly, creates a listener per Kubernetes service.
+
+More broadly, large scale gateway users often expose O(1000) domains, but are currently limited by the maximum of 16 `listeners`.
+
+The spec currently has language to indicate implementations `MAY` merge Gateways resources but the mechanic isn't defined.
+https://github.com/kubernetes-sigs/gateway-api/blob/541e9fc2b3c2f62915cb58dc0ee5e43e4096b3e2/apis/v1beta1/gateway_types.go#L76-L78
+
+## API
+
+This proposal has two aspects: configuration for the parent Gateway, and configuration for the child Gateway.
+
+A "parent" Gateway _does not_ reference another Gateway. A parent Gateway MUST explicitly opt into merging.
+This is done with a new field `allowedChildren` in the `spec.infrastructure` of a Gateway.
+If, and only if, they have done so, the same Gateway is also permitted to leave `listeners` empty (currently, there is a `MinItems=1` restriction).
+
+A "child" Gateway references a parent Gateway. 
+This is done with a new field `attachTo` in the `spec.infrastructure` stanza of a Gateway.
+The `attachTo` field is a new type `GatewayObjectReference`.
+Although the use of `GatewayObjectReference` allows users to attach to any `kind`, this GEP only defines the behavior of attaching a Gateway to another Gateway.
```

Comment:
`sectionName` and `port` don't really mean anything in this context

Reply:
We have a bunch of these here actually - https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1/object_reference_types.go


Big thing is to have smart defaulting - eg. in this scenario we want the kind/type to be a Gateway.

Reply:
I think the general advice in k8s apis is to not excessively reuse types when they are not the same thing. Past example was Ingress and Service shared a type ofr status and a field accidentally got added to both thinking it was just one (oops!).

In this case, there are _already_ irrelevant fields so I think its best to keep it a new type

Reply:
In that case, would a more generalized name like `AttachmentReference` (still defaulting to `group: gateway.networking.k8s.io` and `kind: Gateway`) feel a bit less awkward if it were to ever be used for attachment to non-Gateway kinds?

Reply:
The reason why the type has a `Gateway` prefix wasn't to imply that it only attaches to Gateways but that it's the Gateway that wants to attach to something else.

I'm flexible here - I can see how it can be read as "Attach to this other Gateway"

If we go for `AttachmentReference` then I'd probably drop the defaulting of the gateway kind. This would open up the type to be re-used in the future. eg. HTTPRoute attaching to another one.

Getting more input here would be useful

Reply:
I agree that this should be `GatewayObjectReference` or maybe `ParentGatewayReference` to make the purpose super clear.

Reply:
FWIW I'm mostly thinking of future needs here related to attaching to static infrastructure (#1687), where attachment to a non-Gateway resource such as an Azure [ApplicationLoadBalancer](https://learn.microsoft.com/en-us/azure/application-gateway/for-containers/quickstart-create-application-gateway-for-containers-managed-by-alb-controller?tabs=new-subnet-aks-vnet) should be possible on this field.

Reply:
I like `ParentGatewayReference` because it doesn't imply the Gateway is the _only_  type of object. It would allow for other resources like an Azure ALB.